### ISSUE TITLE
`nix-env --upgrade` upgrades 'nix-2.3.10' to 'nix-2.3.10-x86_64-unknown-linux-musl'

### ISSUE THREAD

**Issue by fishy** [NixOS/nixpkgs#118481]
**Describe the bug**
I'm running nixpkgs on Debian testing on amd64. After doing `nix-channel --update && nix-env --upgrade` it upgraded 'nix-2.3.10' to 'nix-2.3.10-x86_64-unknown-linux-musl'.

After that upgrade is done, `nix-env --upgrade --show-trace` complains about `attribute 'unsafeDiscardStringContext' missing`:

```
$ nix-env --upgrade --show-trace
error: while evaluating anonymous function at /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/pkgs/top-level/impure.nix:15:1, called from undefined position:
while evaluating anonymous function at /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/pkgs/top-level/default.nix:19:1, called from /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/pkgs/top-level/impure.nix:83:1:
while evaluating anonymous function at /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/pkgs/stdenv/booter.nix:42:1, called from /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/pkgs/top-level/default.nix:122:10:
while evaluating 'dfold' at /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/pkgs/stdenv/booter.nix:60:27, called from /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/pkgs/stdenv/booter.nix:136:4:
while evaluating 'go' at /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/pkgs/stdenv/booter.nix:63:18, called from /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/pkgs/stdenv/booter.nix:72:13:
while evaluating 'imap1' at /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/lib/lists.nix:116:14, called from /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/pkgs/stdenv/booter.nix:78:30:
while evaluating 'reverseList' at /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/lib/lists.nix:393:17, called from /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/pkgs/stdenv/booter.nix:85:6:
while evaluating anonymous function at /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/pkgs/stdenv/default.nix:7:1, called from /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/pkgs/top-level/default.nix:118:12:
while evaluating 'elaborate' at /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/lib/systems/default.nix:18:15, called from /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/pkgs/top-level/default.nix:52:17:
while evaluating 'foldl' at /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/lib/lists.nix:80:20, called from /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/lib/systems/default.nix:149:13:
while evaluating 'foldl'' at /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/lib/lists.nix:82:16, called from /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/lib/lists.nix:86:8:
while evaluating the attribute 'parsed.abi.assertions' at /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/lib/systems/default.nix:23:7:
while evaluating 'mkSystemFromString' at /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/lib/systems/parse.nix:462:24, called from /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/lib/systems/default.nix:23:16:
while evaluating 'mkSkeletonFromList' at /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/lib/systems/parse.nix:382:24, called from /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/lib/systems/parse.nix:462:49:
while evaluating 'splitString' at /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/lib/strings.nix:427:23, called from /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/lib/systems/parse.nix:462:69:
attribute 'unsafeDiscardStringContext' missing, at /nix/store/jfqj09iqymfjy7l9z1lk5pbslww8nvl0-nixpkgs-21.05pre280331.54c1e44240d/nixpkgs/lib/strings.nix:429:13
```

**To Reproduce**
1. ...
2. ...
3. ...

**Expected behavior**

**Screenshots**

**Additional context**

**Notify maintainers**

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result. (NOTE: this is after I did `nix-env --rollback`)

 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.0-5-amd64, Debian GNU/Linux, noversion`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - channels(fishy): `"nixpkgs-21.05pre280331.54c1e44240d"`
 - nixpkgs: `/home/fishy/.nix-defexpr/channels/nixpkgs`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by symphorien** [comment]
nix-env --upgrade upgrading packages to different packages because their name looks like a version bump is well known and not really fixable. cc https://github.com/NixOS/nixos-homepage/issues/672
My opinionated advice: use home-manager.

The fact that nix-2.3.10-x86_64-unknown-linux-musl is broken may be viewed as a separate issue deserving further investigation.

**Comment by Ericson2314** [comment]
Definitely agree there is a larger issue here, but why on earth is the cross target coming *after* the version?! That is a big somewhere.

**Comment by edolstra** [comment]
The issue here is that the `nixStatic` package has a name `nix-2.3.10-x86_64-unknown-linux-musl`. This should be `nix-static-2.3.10`. The `x86_64-unknown-linux-musl` part is unhelpful to the user, and in any case features should be part of the `pname`, not the `version`.

**Comment by domenkozar** [comment]
This is fixed in https://github.com/NixOS/nixpkgs/pull/119317

Check https://status.nixos.org/ for when it will reach the nixpkgs-unstable channel!

**Comment by sternenseemann** [comment]
<https://nixpk.gs/pr-tracker.html?pr=119317>

**Comment by DianaOlympos** [comment]
I would like to reopen this and pin it, as it may affect a lot of people and break their install.

The fix was never release to unstable because hydra is broken on unstable. And a lot of people run unstable locally on their dev env.

**Comment by domenkozar** [comment]
I'm working on unblocking the channel.

**Comment by domenkozar** [comment]
Done

**Comment by magthe** [comment]
Are there any instructions on how to fix it on a local install?

My `nix-env --upgrade` is still erroring out with 

```
error: attribute 'unsafeDiscardStringContext' missing, at /nix/store/swryhbsm4cfwfr61g67gmladvigmpw68-nixpkgs-src/lib/strings.nix:730:5
(use '--show-trace' to show detailed location information)
```

**Comment by sternenseemann** [comment]
I think people had success with using `nix-env --rollback` until they had a matching nix-daemon and nix version again.

**Comment by domenkozar** [comment]
You need to update nixpkgs, if you're using channels that's `nix-channel --update`. 

If you're using niv see https://nix.dev/tutorials/towards-reproducibility-pinning-nixpkgs.html

**Comment by mgudemann** [comment]
Is there another solution apart from `nix-env --rollback`? I don't seem to have an old-enough generation available.

**Comment by domenkozar** [comment]
Assuming you're on linux you could try:

```
$(nix-store -r /nix/store/frrxf7qwl5dixa9p591wx9029pv8v066-nix-2.3.10)/bin/nix-env --upgrade

**Comment by mgudemann** [comment]
@domenkozar thanks, this helped

**Comment by a-m-joseph** [comment]
I am still encountering this issue with cross-compiled copies of nix.  This is necessary for the bootstrap process on mips64el.

```
$ nix-build /nix/nixpkgs/ -A pkgsCross.mips64el-linux-gnuabi64.nixStatic
/nix/store/la7rzvxm7wgwqjqz397lrmbgbkwfl653-nix-static-mips64el-unknown-linux-musl-2.3.16
```
and then on the target machine:

```
$ /nix/store/la7rzvxm7wgwqjqz397lrmbgbkwfl653-nix-static-mips64el-unknown-linux-musl-2.3.16/bin/nix-env -iA hello
error: attribute 'unsafeDiscardStringContext' missing, at /nix/nixpkgs/lib/strings.nix:442:13
(use '--show-trace' to show detailed location information)
```

Can anybody enlighten me as to why the name of the directory in which the nix binary lives causes this error?

Is there a better way to use nixpkgs to cross-compile a standalone nix binary that will work on a system that does not yet have a `/nix/store`?

In the past I have used the Debian compiler/assembler/libraries to compile nix from source code, then used that to bootstrap.  That's how I originally got mips64el working.  However that process was tricky and not very repeatable, and as part of the upstreaming process I'd really like to have a documented bootstrap process that starts by using an amd64 machine to cross-compile the nix binary that will be used for bootstrapping.  I'm having quite a difficult time doing that.

**Comment by sternenseemann** [comment]
I think in your case the issue is rather that what nix assumes to be `<nixpkgs>` doesn't exist, namely the `/nix/nixpkgs` directory, which is not a special `/nix` path, but rather has been created by you manually on your other machine, I assume. Seems that maybe `NIX_PATH` leaked into the mips machine or the nix binary.

**Comment by a-m-joseph** [comment]
> I think in your case the issue is rather that what nix assumes to be `<nixpkgs>` doesn't exist, namely the `/nix/nixpkgs` directory,

Unfortunately that is not the issue; I copied (git clone) /nix/nixpkgs from the machine which built nix to the machine on which I'm running it.

```
$ ls /nix/nixpkgs/lib/strings.nix
/nix/nixpkgs/lib/strings.nix
```

**Comment by a-m-joseph** [comment]
I was able to work around this by simply using nix 2.4 instead, in case that helps anybody else.

```
nix-build -A pkgsCross.mips64el-linux-gnuabi64.pkgsStatic.nix_2_4
```
